Store and Manage permissions in the Registry

tmihalac opened the pull request:
Adding management of Permissions in registry

dmartinol committed:
Introducing permission framework and authorization manager in user guide (to be continued after the code is consolidated)

Signed-off-by: Daniele Martinoli <86618610+dmartinol@users.noreply.github.com>

dmartinol committed:
removed test code

Signed-off-by: Daniele Martinoli <86618610+dmartinol@users.noreply.github.com>

dmartinol committed:
hiding sensitive data (false positive, anyway)

Signed-off-by: Daniele Martinoli <86618610+dmartinol@users.noreply.github.com>

tmihalac committed:
Store and Manage permissions in the Registry

Signed-off-by: Theodor Mihalache <tmihalac@redhat.com>

tmihalac committed:
Store and Manage permissions in the Registry

Signed-off-by: Theodor Mihalache <tmihalac@redhat.com>

redhatHameed commented on protos/feast/core/Permission.proto:
```diff
@@ -0,0 +1,57 @@
+
+  // the policy.
+  Policy policy = 8;
+}
```
can you add new/empty line here

tmihalac replied:
you mean end of file ? or between curly braces and policy line ?

redhatHameed replied:
end of file to remove that sign 
<img width="569" alt="Screenshot 2024-06-28 at 10 12 33 AM" src="https://github.com/RHEcosystemAppEng/feast/assets/41960152/404bea69-5126-47f0-8f1c-2f201c6ae610">

redhatHameed commented on protos/feast/core/Policy.proto:
```diff
@@ -0,0 +1,23 @@
+message RoleBasedPolicy {
+  // List of roles in this policy.
+  repeated string roles = 1;
+}
```
add empty line here

redhatHameed commented on protos/feast/registry/RegistryServer.proto:
```diff
@@ -316,4 +323,33 @@ message DeleteValidationReferenceRequest {
+  string name = 1;
+  string project = 2;
+  bool commit = 3;
 }
```
same here add new line

redhatHameed commented on sdk/python/feast/feature_store.py:
```diff
@@ -693,99 +694,6 @@ def _get_feature_views_to_materialize(
 
         return feature_views_to_materialize
 
-    def plan(
```
any reason for changing the position for this function

tmihalac replied:
By mistake

redhatHameed reviewed (commented):
looks good to me, just minor comments

tmihalac committed:
Applied review comments

Signed-off-by: Theodor Mihalache <tmihalac@redhat.com>

tmihalac committed:
Applied review comments

Signed-off-by: Theodor Mihalache <tmihalac@redhat.com>

tmihalac committed:
Store and Manage permissions in the Registry
- Fixes to code
- Made test case broader

Signed-off-by: Theodor Mihalache <tmihalac@redhat.com>

tmihalac committed:
Store and Manage permissions in the Registry
- Fixes to code
- Made test case broader

Signed-off-by: Theodor Mihalache <tmihalac@redhat.com>

tmihalac committed:
Store and Manage permissions in the Registry
- Fixed incorrectly recognized linter error

Signed-off-by: Theodor Mihalache <tmihalac@redhat.com>

tmihalac committed:
Store and Manage permissions in the Registry
- Fixed incorrectly recognized linter error

Signed-off-by: Theodor Mihalache <tmihalac@redhat.com>

dmartinol commented on sdk/python/feast/diff/registry_diff.py:
```diff
@@ -129,8 +132,10 @@ def diff_registry_objects(
     if isinstance(
-        current_proto, (DataSourceProto, ValidationReferenceProto)
-    ) or isinstance(new_proto, (DataSourceProto, ValidationReferenceProto)):
+        current_proto, (DataSourceProto, ValidationReferenceProto, PermissionProto)
```
are you sure we need this casting for Permissions? (just asking, I imagine this is done to handle sub-classes, as DataSource is abstract, but not completely sure about that)

tmihalac replied:
Not sure consulting with Tornike regarding the exisiting code, waiting for his input on this, will mark it as resolved for now, will change it in the future if its needed

dmartinol commented on sdk/python/feast/feast_object.py:
```diff
@@ -41,4 +44,7 @@
+    PermissionProto,
 ]
+
+ALL_RESOURCE_TYPES = list(get_args(FeastObject))
```
maybe we can rename it to `ALL_FEAST_OBJECTS` since we moved it to the root module?

tmihalac replied:
Is it really objects? isn't it object types ?

dmartinol replied:
Yes, but `FeastObject` is the type name 😄  Anyway, any name is fine

tmihalac replied:
I think the current name is more informative, unless its really important for you I will leave it be

dmartinol commented on sdk/python/feast/permissions/permission.py:
```diff
@@ -137,6 +161,68 @@ def match_actions(self, actions: list[AuthzedAction]) -> bool:
+        """
+        Converts a PermissionProto object to its protobuf representation.
+        """
+        types = [
```
can't you use the `_PERMISSION_TYPES` dictionary instead? it should already have the same mapping available with no pain

tmihalac replied:
no, I need a list of ValueType objects and in the _PERMISSION_TYPES dictionary I have strings

dmartinol replied:
Something like:
```py
get_type_class_from_permission_type(
                _PERMISSION_TYPES[config_class_name]
            )
```

tmihalac replied:
I am using it where I can, see in `from_proto` but in this case I can't, I need `ValueType` object from the `proto` and not the actual feast object type

dmartinol commented on sdk/python/feast/permissions/permission.py:
```diff
@@ -151,3 +237,22 @@ def _normalize_required_tags(required_tags: Optional[dict[str, str]]):
+    return import_class(module_name, config_class_name)
+
+
+_PERMISSION_TYPES = {
```
Shouldn't we also move this map to the feast_object module?
And maybe also the functions to convert the string to the type and vice versa

tmihalac replied:
I am not sure, as all the things you mentioned are related to permissions, if its really important for you I will move them

dmartinol commented on sdk/python/feast/registry_server.py:
```diff
@@ -336,6 +337,47 @@ def GetInfra(self, request: RegistryServer_pb2.GetInfraRequest, context):
+
+    def ListPermissions(
+        self, request: RegistryServer_pb2.ListPermissionsRequest, context
+    ):
+        permissions_proto = []
+        permissions = self.proxied_registry.list_permissions(
+            project=request.project,
+            allow_cache=request.allow_cache,
+        )
+        for p in permissions:
+            permission_proto = p.to_proto()
+            permission_proto.project = request.project
+            permissions_proto.append(permission_proto)
+
+        return RegistryServer_pb2.ListPermissionsResponse(permissions=permissions_proto)
```
Why don't we use the regular pattern:
```py
        return RegistryServer_pb2.ListPermissionsResponse(
            permissions=[
                permission.to_proto()
                for permission in self.proxied_registry.list_permissions(
                    project=request.project,
                    allow_cache=request.allow_cache,
                )
            ]
        )
```

BTW: do we really need to set the `project` field? I can't see that this is managed in the equivalent methods for other resource types

tmihalac replied:
You are right, I just didn't want to take any chances, will remove it

dmartinol commented on sdk/python/feast/repo_operations.py:
```diff
@@ -201,6 +203,10 @@ def parse_repo(repo_root: Path) -> RepoContents:
             ):
                 res.on_demand_feature_views.append(obj)
+            elif isinstance(obj, Permission) and not any(
+                (obj is fs) for fs in res.permissions
```
maybe `fs` is misleading here

tmihalac replied:
You are right

dmartinol reviewed (commented):
lgtm

dmartinol committed:
Merge pull request #23 from dmartinol/feast-rbac

User guide

dmartinol committed:
Added filter_only flag to assert_permissions and returning a list of filtered resources instead of PermissionError

Signed-off-by: Daniele Martinoli <86618610+dmartinol@users.noreply.github.com>

dmartinol committed:
added the option to return the single resource, or None

Signed-off-by: Daniele Martinoli <86618610+dmartinol@users.noreply.github.com>

redhatHameed committed:
Merge pull request #26 from dmartinol/feast-rbac

Added filter_only flag to assert_permissions

tmihalac committed:
Store and Manage permissions in the Registry
- Added test
- Fixed missing property to permission
- Changed code following review

Signed-off-by: Theodor Mihalache <tmihalac@redhat.com>

tmihalac committed:
Merge remote-tracking branch 'origin/store-permission-in-registry' into store-permission-in-registry

# Conflicts:
#	protos/feast/core/Permission.proto
#	sdk/python/feast/permissions/permission.py
#	sdk/python/feast/registry_server.py
#	sdk/python/feast/repo_operations.py
